privacyidea/lib/tokens/pushtoken.py

```diff
@@ -0,0 +1,582 @@
+
+        :param transactionid: the id of this challenge
+        :param options: the request context parameters / data
+        :type options: dict
+        :return: tuple of (bool, message, transactionid, attributes)
+        :rtype: tuple
+
+        The return tuple builds up like this:
+        ``bool`` if submit was successful;
+        ``message`` which is displayed in the JSON response;
+        additional ``attributes``, which are displayed in the JSON response.
+        """
+        res = False
+        options = options or {}
+        message = get_action_values_from_options(SCOPE.AUTH,
+                                                 ACTION.CHALLENGETEXT,
+                                                 options) or DEFAULT_CHALLENGE_TEXT
+
+        attributes = None
+        data = None
+        challenge = b32encode_and_unicode(geturandom())
+        fb_identifier = self.get_tokeninfo(PUSH_ACTION.FIREBASE_CONFIG)
+        if fb_identifier:
+            # We send the challenge to the Firebase service
+            fb_gateway = create_sms_instance(fb_identifier)
+            url = fb_gateway.smsgateway.option_dict.get(FIREBASE_CONFIG.REGISTRATION_URL)
+            message_on_mobile = get_action_values_from_options(SCOPE.AUTH,
+                                                   PUSH_ACTION.MOBILE_TITLE,
+                                                   options) or DEFAULT_MOBILE_TEXT
+            title = get_action_values_from_options(SCOPE.AUTH,
+                                                   PUSH_ACTION.MOBILE_TITLE,
+                                                   options) or "privacyIDEA"
+            smartphone_data = {"nonce": challenge,
+                               "question": message_on_mobile,
+                               "serial": self.token.serial,
+                               "title": title,
+                               "url": url}
+            # Create the signature.
+            # value to string
+            sign_string = u"{nonce}|{url}|{serial}|{question}|{title}".format(**smartphone_data)
+
+            pem_privkey = self.get_tokeninfo(PRIVATE_KEY_SERVER)
+            privkey_obj = serialization.load_pem_private_key(to_bytes(pem_privkey), None, default_backend())
+
+            # Sign the data with PKCS1 padding. Not all Androids support PSS padding.
+            signature = privkey_obj.sign(sign_string.encode("utf8"),
+                                         padding.PKCS1v15(),
+                                         hashes.SHA256())
+            smartphone_data["signature"] = b32encode_and_unicode(signature)
+
+            res = fb_gateway.submit_message(self.get_tokeninfo("firebase_token"), smartphone_data)
+            if not res:
+                from privacyidea.lib.error import ValidateError
+                raise ValidateError("Failed to submit message to firebase service.")
+        else:
+            log.warning(u"The token {0!s} has no tokeninfo {1!s}. "
```

Comment:
Should we maybe exit from the function in this case?

Reply:
done in 7544a675